oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,358 @@
+   * @return the {@link AwsRequestSignature}.
+   */
+  AwsRequestSignature sign() {
+    // Get the dates to be used to sign the request.
+    AwsDates dates = getDates();
+
+    // Retrieve the service name. e.g. iam.amazonaws.com host => iam service.
+    String serviceName = Splitter.on(".").split(uri.getHost()).iterator().next();
+
+    Map<String, String> canonicalHeaders = getCanonicalHeaders(dates.getOriginalDate());
+    // Headers must be sorted.
+    List<String> sortedHeaders = new ArrayList<>();
+    for (String headerName : canonicalHeaders.keySet()) {
+      sortedHeaders.add(headerName.toLowerCase());
+    }
+    Collections.sort(sortedHeaders);
+
+    String canonicalRequestHash = createCanonicalRequestHash(canonicalHeaders, sortedHeaders);
+    String credentialScope =
+        dates.getFormattedDate() + "/" + region + "/" + serviceName + "/" + AWS_REQUEST_TYPE;
+    String stringToSign =
+        createStringToSign(canonicalRequestHash, dates.getAmzDate(), credentialScope);
+    String signature =
+        calculateAwsV4Signature(
+            serviceName,
+            awsSecurityCredentials.getSecretAccessKey(),
+            dates.getFormattedDate(),
+            region,
+            stringToSign);
+
+    return new AwsRequestSignature.Builder()
+        .setSignature(signature)
+        .setCanonicalHeaders(canonicalHeaders)
+        .setHttpMethod(httpMethod)
+        .setSecurityCredentials(awsSecurityCredentials)
+        .setCredentialScope(credentialScope)
+        .setUrl(uri.toString())
+        .setDate(dates.getOriginalDate())
+        .setSortedHeaderNames(sortedHeaders)
+        .setRegion(region)
+        .build();
+  }
+
+  /** Task 1: Create a canonical request for Signature Version 4. */
+  private String createCanonicalRequestHash(
+      Map<String, String> headers, List<String> sortedHeaderNames) {
+    // Append the HTTP request method.
+    StringBuilder canonicalRequest = new StringBuilder(httpMethod).append("\n");
+
+    // Append the path.
+    String urlPath = uri.getRawPath().isEmpty() ? "/" : uri.getRawPath();
+    canonicalRequest.append(urlPath).append("\n");
+
+    // Append the canonical query string.
+    String actionQueryString = uri.getRawQuery() != null ? uri.getRawQuery() : "";
+    canonicalRequest.append(actionQueryString).append("\n");
+
+    // Append the canonical headers.
+    StringBuilder canonicalHeaders = new StringBuilder();
+    for (String headerName : sortedHeaderNames) {
+      canonicalHeaders.append(headerName).append(":").append(headers.get(headerName)).append("\n");
+    }
+    canonicalRequest.append(canonicalHeaders).append("\n");
+
+    // Append the signed headers.
+    canonicalRequest.append(Joiner.on(';').join(sortedHeaderNames)).append("\n");
+
+    // Append the hashed request payload.
+    canonicalRequest.append(DigestUtils.sha256Hex(requestPayload));
+
+    // Return the hashed canonical request.
+    return DigestUtils.sha256Hex(canonicalRequest.toString());
```

Comment:
Let's switch to guava's sha256 digester so we don't need to use commons-codec

Reply:
Actually, let's go with [MessageDigest](https://docs.oracle.com/javase/7/docs/api/java/security/MessageDigest.html)